File: platform/android-customer/app/src/main/java/ca/shawcliffe/tomsproduce/StorefrontViewModel.kt
package ca.shawcliffe.tomsproduce

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.github.jan.supabase.postgrest.from
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

class StorefrontViewModel : ViewModel() {
    var branding by mutableStateOf<ClientBranding?>(null)
        private set
    var status by mutableStateOf<DailyStatus?>(null)
        private set
    var location by mutableStateOf<Location?>(null)
        private set
    var products by mutableStateOf<List<Product>>(emptyList())
        private set
    var isLoading by mutableStateOf(true)
        private set
    var errorMessage by mutableStateOf<String?>(null)
        private set

    private val clientId = Config.CLIENT_ID
    private var pollJob: Job? = null
    private val today: String = SimpleDateFormat("yyyy-MM-dd", Locale.US).apply {
        timeZone = TimeZone.getDefault()
    }.format(Date())

    fun start() {
        viewModelScope.launch {
            loadAll()
            isLoading = false
            startPolling()
        }
    }

    fun stop() {
        pollJob?.cancel()
        pollJob = null
    }

    private fun startPolling() {
        pollJob?.cancel()
        pollJob = viewModelScope.launch {
            while (isActive) {
                delay(5_000)
                if (!isActive) break
                loadStatus()
                loadProducts()
            }
        }
    }

    private suspend fun loadAll() {
        loadBranding()
        loadStatus()
        loadProducts()
    }

    private suspend fun loadBranding() {
        try {
            branding = supabase.from("client_branding").select {
                filter { eq("client_id", clientId) }
            }.decodeList<ClientBranding>().firstOrNull()
        } catch (e: Exception) {
            errorMessage = "Couldn't load storefront."
        }
    }

    private suspend fun loadStatus() {
        try {
            val rows = supabase.from("daily_status").select {
                filter {
                    eq("client_id", clientId)
                    eq("date", today)
                }
            }.decodeList<DailyStatus>()
            status = rows.firstOrNull()
            val locationId = status?.locationId
            if (locationId != null) {
                loadLocation(locationId)
            } else {
                location = null
            }
        } catch (e: Exception) {
            // Keep showing last-known status rather than clobbering it with an error.
        }
    }

    private suspend fun loadLocation(id: String) {
        try {
            location = supabase.from("locations").select {
                filter {
                    eq("client_id", clientId)
                    eq("id", id)
                }
            }.decodeList<Location>().firstOrNull()
        } catch (e: Exception) {
            // Non-critical — leave last-known location in place.
        }
    }

    private suspend fun loadProducts() {
        try {
            products = supabase.from("products").select {
                filter { eq("client_id", clientId) }
                order("sort_order", io.github.jan.supabase.postgrest.query.Order.ASCENDING)
            }.decodeList()
        } catch (e: Exception) {
            // Non-critical — leave last-known products in place.
        }
    }
}

File: platform/ios-customer/ShawcliffeCustomer/StorefrontViewModel.swift
import Foundation

@MainActor
final class StorefrontViewModel: ObservableObject {
    @Published var branding: ClientBranding?
    @Published var status: DailyStatus?
    @Published var location: Location?
    @Published var products: [Product] = []
    @Published var enabledComponents: [String] = []
    @Published var isLoading = true
    @Published var errorMessage: String?

    private let clientId = Config.clientId
    private var pollTask: Task<Void, Never>?
    private let today: String = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.timeZone = .current
        return formatter.string(from: Date())
    }()

    func start() async {
        await loadAll()
        isLoading = false
        startPolling()
    }

    func stop() {
        pollTask?.cancel()
        pollTask = nil
    }

    private func startPolling() {
        pollTask?.cancel()
        pollTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 5_000_000_000)
                if Task.isCancelled { break }
                await loadStatus()
                await loadProducts()
            }
        }
    }

    private func loadAll() async {
        async let brandingTask: Void = loadBranding()
        async let statusTask: Void = loadStatus()
        async let productsTask: Void = loadProducts()
        async let componentsTask: Void = loadComponents()
        _ = await (brandingTask, statusTask, productsTask, componentsTask)
    }

    private func loadComponents() async {
        do {
            let url = Config.apiBaseURL.appendingPathComponent("api/client/\(clientId)/components")
            let (data, _) = try await URLSession.shared.data(from: url)
            enabledComponents = try JSONDecoder().decode(EnabledComponentsResponse.self, from: data).enabled_components
        } catch {
            // Fail closed — status tracker just stays hidden.
            enabledComponents = []
            print("Failed to load enabled components: \(error)")
        }
    }

    private func loadBranding() async {
        do {
            let rows: [ClientBranding] = try await supabase
                .from("client_branding")
                .select()
                .eq("client_id", value: clientId)
                .execute()
                .value
            branding = rows.first
        } catch {
            errorMessage = "Couldn't load storefront."
        }
    }

    private func loadStatus() async {
        do {
            let rows: [DailyStatus] = try await supabase
                .from("daily_status")
                .select()
                .eq("client_id", value: clientId)
                .eq("date", value: today)
                .execute()
                .value
            status = rows.first
            if let locationId = status?.locationId {
                await loadLocation(id: locationId)
            } else {
                location = nil
            }
        } catch {
            // Keep showing last-known status rather than clobbering it with an error.
        }
    }

    private func loadLocation(id: String) async {
        do {
            let rows: [Location] = try await supabase
                .from("locations")
                .select()
                .eq("client_id", value: clientId)
                .eq("id", value: id)
                .execute()
                .value
            location = rows.first
        } catch {
            // Non-critical — leave last-known location in place.
        }
    }

    private func loadProducts() async {
        do {
            let rows: [Product] = try await supabase
                .from("products")
                .select()
                .eq("client_id", value: clientId)
                .order("sort_order")
                .execute()
                .value
            products = rows
        } catch {
            // Non-critical — leave last-known products in place.
        }
    }
}
